db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/UnaryComparisonOperatorNode.java

```diff
@@ -390,7 +391,15 @@ public boolean isQualifier(Optimizable optTable, boolean forPush) {
         ColumnReference cr = (ColumnReference) getOperand();
         FromTable ft = (FromTable) optTable;
 
-        return cr.getTableNumber()==ft.getTableNumber();
+        if (cr.getTableNumber() != ft.getTableNumber()) {
+            return false;
+        }
+
+        if (forPush && cd != null && cd.isIndex()) {
```

Comment:
It seems like we'd want to do this check even when forPush is false so that higher level calls like isIndexUseful are correct and we can make the correct costing and join planning decisions.

Reply:
This is explained in the comment of `isQualifier()` in `RelationalOperator.java`. When `forPush` is `false`, this method is used to check if this predicate can be used as a hash key. In that case, we should not check for index columns because the predicate will be on top of index lookup (if the index is not covering). If we check for index columns, we may miss the chance of using this predicate as a hash key, thus disabling hashable join.

`isIndexUseful()` logic is correct even before this fix because it uses `isQualifier()` as only one of its preconditions. If a predicate is not a qualifier, it rejects the predicate. If it is, `isIndexUseful()` further checks for index columns for both classic index and index on expressions cases. So it should be fine.

The strange part, though, is that we have too many methods essentially check for the same thing: `isIndexUseful()`, `isQualifier()`, `isHashJoinQualifier()`, `isPushableOrPredicate()`, etc. ...IMHO this part should be carefully refactored.

Reply:
OK, I see, thanks.  Yes, as part of DB-11449 I needed to add a special pushOnlyHashableJoinPreds flag to orderUsefulPredicates.  I found it strange that we have to push join predicates to the base table in order to find them for hashable joins, even if they aren't qualifiers.  It seems like we should allow the predicates to be pushed just to the ProjectRestrict above the FromBaseTable, and also look for hashable join predicates there.  I agree some it would be good to refactor this and maybe modify the current predicate pushing behavior.